File: src/agentrust_telemetry/propagation.py
"""W3C trace-context propagation with minimal AgentTrust correlation metadata."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, MutableMapping

from .errors import PropagationError


RUN_ID_HEADER = "x-agentrust-run-id"
WORKFLOW_ID_HEADER = "x-agentrust-workflow-id"
UPSTREAM_AGENT_ID_HEADER = "x-agentrust-agent-id"
_MAX_VALUE_LENGTH = 512


@dataclass(frozen=True)
class ExtractedContext:
    """Remote OTel context plus untrusted AgentTrust correlation metadata."""

    otel_context: Any
    run_id: str | None
    workflow_id: str | None
    upstream_agent_id: str | None

    def event_fields(self, *, agent_id: str) -> dict[str, str]:
        """Build envelope fields for work performed by ``agent_id``."""
        fields = {"agent_id": _safe_value("agent_id", agent_id, required=True)}
        if self.run_id is not None:
            fields["run_id"] = self.run_id
        if self.workflow_id is not None:
            fields["workflow_id"] = self.workflow_id
        if self.upstream_agent_id is not None:
            fields["parent_agent_id"] = self.upstream_agent_id
        return fields

    def link(self, *, attributes: dict[str, Any] | None = None) -> Any:
        """Create an OTel Link for asynchronous handoff or fan-in."""
        _, trace = _otel()
        span_context = trace.get_current_span(self.otel_context).get_span_context()
        if not span_context.is_valid:
            raise PropagationError("extracted context has no valid remote span to link")
        return trace.Link(span_context, attributes=attributes)


def inject_context(
    carrier: MutableMapping[str, str],
    *,
    run_id: str,
    agent_id: str,
    workflow_id: str | None = None,
    context: Any | None = None,
) -> None:
    """Inject standard W3C context and minimal AgentTrust headers into ``carrier``."""
    propagate, _ = _otel()
    propagate.inject(carrier, context=context)
    carrier[RUN_ID_HEADER] = _safe_value("run_id", run_id, required=True)
    carrier[UPSTREAM_AGENT_ID_HEADER] = _safe_value("agent_id", agent_id, required=True)
    if workflow_id is not None:
        carrier[WORKFLOW_ID_HEADER] = _safe_value("workflow_id", workflow_id, required=True)


def extract_context(carrier: MutableMapping[str, str]) -> ExtractedContext:
    """Extract a remote parent and validate optional AgentTrust headers."""
    propagate, _ = _otel()
    context = propagate.extract(carrier)
    return ExtractedContext(
        otel_context=context,
        run_id=_header(carrier, RUN_ID_HEADER),
        workflow_id=_header(carrier, WORKFLOW_ID_HEADER),
        upstream_agent_id=_header(carrier, UPSTREAM_AGENT_ID_HEADER),
    )


def _header(carrier: MutableMapping[str, str], name: str) -> str | None:
    value = carrier.get(name)
    if value is None:
        # HTTP field names are case-insensitive even when a plain mapping is used.
        value = next((item for key, item in carrier.items() if key.lower() == name), None)
    return None if value is None else _safe_value(name, value, required=True)


def _safe_value(name: str, value: Any, *, required: bool) -> str:
    if not isinstance(value, str):
        raise PropagationError(f"{name} must be a string")
    if required and not value:
        raise PropagationError(f"{name} must not be empty")
    if len(value) > _MAX_VALUE_LENGTH:
        raise PropagationError(f"{name} exceeds {_MAX_VALUE_LENGTH} characters")
    if "\r" in value or "\n" in value:
        raise PropagationError(f"{name} contains a prohibited line break")
    return value


def _otel() -> tuple[Any, Any]:
    try:
        from opentelemetry import propagate, trace
    except ImportError as exc:  # pragma: no cover - exercised in minimal installs
        raise PropagationError(
            "OpenTelemetry propagation requires the 'otel' optional dependency"
        ) from exc
    return propagate, trace

File: packages/typescript/src/propagation.ts
import {context, isSpanContextValid, propagation, trace, type Context, type Link} from "@opentelemetry/api";

export const RUN_ID_HEADER = "x-agentrust-run-id";
export const WORKFLOW_ID_HEADER = "x-agentrust-workflow-id";
export const UPSTREAM_AGENT_ID_HEADER = "x-agentrust-agent-id";
const MAX = 512;
export class PropagationError extends Error {}

export class ExtractedContext {
  constructor(readonly otelContext: Context, readonly runId?: string, readonly workflowId?: string, readonly upstreamAgentId?: string) {}
  eventFields(agentId: string): Record<string, string> {return {agent_id: safe("agent_id", agentId), ...(this.runId ? {run_id: this.runId} : {}), ...(this.workflowId ? {workflow_id: this.workflowId} : {}), ...(this.upstreamAgentId ? {parent_agent_id: this.upstreamAgentId} : {})};}
  link(attributes?: Record<string, string | number | boolean>): Link {const spanContext = trace.getSpanContext(this.otelContext); if (!spanContext || !isSpanContextValid(spanContext)) throw new PropagationError("extracted context has no valid remote span to link"); return {context: spanContext, ...(attributes ? {attributes} : {})};}
}
export function injectContext(carrier: Record<string, string>, fields: {runId: string; agentId: string; workflowId?: string; otelContext?: Context}): void {
  propagation.inject(fields.otelContext ?? context.active(), carrier);
  carrier[RUN_ID_HEADER] = safe("run_id", fields.runId); carrier[UPSTREAM_AGENT_ID_HEADER] = safe("agent_id", fields.agentId);
  if (fields.workflowId !== undefined) carrier[WORKFLOW_ID_HEADER] = safe("workflow_id", fields.workflowId);
}
export function extractContext(carrier: Record<string, string>): ExtractedContext {
  return new ExtractedContext(propagation.extract(context.active(), carrier, {keys: (c) => Object.keys(c), get: (c, key) => header(c, key)}), value(carrier, RUN_ID_HEADER), value(carrier, WORKFLOW_ID_HEADER), value(carrier, UPSTREAM_AGENT_ID_HEADER));
}
function value(carrier: Record<string, string>, name: string): string | undefined {const found = header(carrier, name); return found === undefined ? undefined : safe(name, Array.isArray(found) ? found[0]! : found);}
function header(carrier: Record<string, string>, name: string): string | string[] | undefined {const key = Object.keys(carrier).find((candidate) => candidate.toLowerCase() === name.toLowerCase()); return key ? carrier[key] : undefined;}
function safe(name: string, value: string): string {if (typeof value !== "string" || !value) throw new PropagationError(`${name} must be a non-empty string`); if (value.length > MAX) throw new PropagationError(`${name} exceeds ${MAX} characters`); if (/\r|\n/.test(value)) throw new PropagationError(`${name} contains a prohibited line break`); return value;}
